In [ ]:
! pip install pandas 

In [3]:
import psycopg2
import pandas as pd
import numpy as np
from sqlalchemy import create_engine, text
import warnings
warnings.filterwarnings("ignore")

In [4]:
DB_URL = "postgresql+psycopg2://postgres:postgres123@localhost:5432/companies_db"
engine = create_engine(DB_URL, connect_args={"gssencmode": "disable"})

print("=" * 60)
print("COMPANIES EDA  —  17M rows")
print("=" * 60)

COMPANIES EDA  —  17M rows


In [5]:
with engine.connect() as con:
    total_rows = con.execute(text("SELECT COUNT(*) FROM companies")).scalar()
    total_cols = con.execute(text(
        "SELECT COUNT(*) FROM information_schema.columns WHERE table_name='companies'"
    )).scalar()

print(f"\nShape: {total_rows:,} rows × {total_cols} columns")



Shape: 17,154,017 rows × 11 columns


In [6]:
CHUNK = 500_000  # rows per chunk

# Columns to cast as category (low cardinality)
CAT_COLS = ["industry", "size", "type", "country_code", "state"]

# Accumulate stats across chunks
null_counts   = {}
value_counts  = {c: pd.Series(dtype="object") for c in CAT_COLS + ["city"]}
founded_stats = []   # collect non-null founded values in a list for numeric EDA
total_seen    = 0

print(f"\nStreaming {total_rows:,} rows in chunks of {CHUNK:,} …")

chunks = []
for i, chunk in enumerate(
    pd.read_sql(
        "SELECT * FROM companies",
        engine,
        chunksize=CHUNK
    )
):
    # ── dtype optimisation ──────────────────────────────────
    for col in CAT_COLS:
        if col in chunk.columns:
            chunk[col] = chunk[col].astype("category")

    # ── null counts ─────────────────────────────────────────
    for col in chunk.columns:
        null_counts[col] = null_counts.get(col, 0) + chunk[col].isna().sum()

    # ── value-count accumulators ────────────────────────────
    for col in value_counts:
        if col in chunk.columns:
            vc = chunk[col].value_counts(dropna=False)
            value_counts[col] = value_counts[col].add(vc, fill_value=0)

    # ── numeric: founded ────────────────────────────────────
    if "founded" in chunk.columns:
        founded_stats.extend(chunk["founded"].dropna().astype(int).tolist())

    total_seen += len(chunk)
    chunks.append(chunk)
    print(f"  chunk {i+1:>3}  ({total_seen:>12,} / {total_rows:,} rows)", end="\r")

print(f"\n✓ All {total_seen:,} rows loaded.")

# Concat for row-level ops (uses category dtypes → memory efficient)
df = pd.concat(chunks, ignore_index=True)
for col in CAT_COLS:
    if col in df.columns:
        df[col] = df[col].astype("category")

print(f"\nMemory usage: {df.memory_usage(deep=True).sum() / 1e9:.2f} GB")



Streaming 17,154,017 rows in chunks of 500,000 …
  chunk  35  (  17,154,017 / 17,154,017 rows)
✓ All 17,154,017 rows loaded.

Memory usage: 5.62 GB


In [7]:
print("\n" + "─" * 60)
print("MISSING VALUES")
print("─" * 60)
mv = pd.DataFrame({
    "nulls":   null_counts,
    "pct_null": {c: round(v / total_rows * 100, 2) for c, v in null_counts.items()}
}).sort_values("pct_null", ascending=False)
print(mv.to_string())


────────────────────────────────────────────────────────────
MISSING VALUES
────────────────────────────────────────────────────────────
                 nulls  pct_null
domain        17154017    100.00
founded        8650132     50.43
type           5575557     32.50
state          4958934     28.91
city           3599298     20.98
website        3511965     20.47
country_code   3140804     18.31
size           2867331     16.72
industry       1309256      7.63
name              4683      0.03
handle               0      0.00


────────────────────────────────────────────────────────────
MISSING VALUES
────────────────────────────────────────────────────────────
                 nulls  pct_null
domain        17154017    100.00
founded        8650132     50.43
type           5575557     32.50
state          4958934     28.91
city           3599298     20.98
website        3511965     20.47
country_code   3140804     18.31
size           2867331     16.72
industry       1309256      7.63
name              4683      0.03
handle               0      0.00

In [2]:
mv

mv: missing file operand
Try 'mv --help' for more information.


In [8]:
print("\n" + "─" * 60)
print("CARDINALITY  (unique values per column)")
print("─" * 60)
for col in df.columns:
    n_unique = df[col].nunique()
    print(f"  {col:<15} {n_unique:>10,} unique")


────────────────────────────────────────────────────────────
CARDINALITY  (unique values per column)
────────────────────────────────────────────────────────────
  handle          17,154,016 unique
  name            16,695,991 unique
  domain                   0 unique
  website         12,054,468 unique
  industry               542 unique
  size                     8 unique
  type                     8 unique
  founded                802 unique
  city               393,449 unique
  state               39,887 unique
  country_code           272 unique


────────────────────────────────────────────────────────────
CARDINALITY  (unique values per column)
────────────────────────────────────────────────────────────
  handle          17,154,016 unique
  name            16,695,991 unique
  domain                   0 unique
  website         12,054,468 unique
  industry               542 unique
  size                     8 unique
  type                     8 unique
  founded                802 unique
  city               393,449 unique
  state               39,887 unique
  country_code           272 unique

In [9]:
for col in CAT_COLS:
    if col not in df.columns:
        continue
    vc = value_counts[col].sort_values(ascending=False)
    top = vc.head(15).astype(float)
    print(f"\n{'─'*60}")
    print(f"TOP 15  {col.upper()}  (total unique: {len(vc):,})")
    print("─" * 60)
    pct = (top / total_rows * 100).round(2)
    summary = pd.DataFrame({"count": top.astype(int), "pct%": pct})
    print(summary.to_string())



────────────────────────────────────────────────────────────
TOP 15  INDUSTRY  (total unique: 543)
────────────────────────────────────────────────────────────
                                        count  pct%
industry                                           
NaN                                   1309256  7.63
construction                           683107  3.98
it services and it consulting          651987  3.80
advertising services                   617887  3.60
business consulting and services       508608  2.96
real estate                            471242  2.75
retail                                 436774  2.55
software development                   406244  2.37
financial services                     356450  2.08
hospitals and health care              303721  1.77
technology, information and internet   297403  1.73
wellness and fitness services          250817  1.46
hospitality                            220073  1.28
non-profit organizations               218232  1.27
profess

────────────────────────────────────────────────────────────
TOP 15  INDUSTRY  (total unique: 543)
────────────────────────────────────────────────────────────
                                        count  pct%
industry                                           
NaN                                   1309256  7.63
construction                           683107  3.98
it services and it consulting          651987  3.80
advertising services                   617887  3.60
business consulting and services       508608  2.96
real estate                            471242  2.75
retail                                 436774  2.55
software development                   406244  2.37
financial services                     356450  2.08
hospitals and health care              303721  1.77
technology, information and internet   297403  1.73
wellness and fitness services          250817  1.46
hospitality                            220073  1.28
non-profit organizations               218232  1.27
professional training and coaching     214756  1.25

────────────────────────────────────────────────────────────
TOP 15  SIZE  (total unique: 9)
────────────────────────────────────────────────────────────
           count   pct%
size                   
1-10     8359109  48.73
11-50    3832365  22.34
NaN      2867331  16.72
51-200   1278419   7.45
201-500   443031   2.58
501-1K    159304   0.93
1K-5K     136315   0.79
10K+       48583   0.28
5K-10K     29560   0.17

────────────────────────────────────────────────────────────
TOP 15  TYPE  (total unique: 9)
────────────────────────────────────────────────────────────
                     count   pct%
type                             
Privately Held     5704843  33.26
NaN                5575557  32.50
Self-Owned         1421913   8.29
Partnership        1135656   6.62
Public Company     1057455   6.16
Self-Employed       931354   5.43
Nonprofit           867050   5.05
Educational         365892   2.13
Government Agency    94297   0.55

────────────────────────────────────────────────────────────
TOP 15  COUNTRY_CODE  (total unique: 273)
────────────────────────────────────────────────────────────
                count   pct%
country_code                
US            4306855  25.11
NaN           3140804  18.31
GB            1294632   7.55
IN             995101   5.80
FR             832545   4.85
BR             719757   4.20
DE             507443   2.96
ES             447529   2.61
NL             434123   2.53
CA             423082   2.47
AU             404922   2.36
IT             349130   2.04
CN             190933   1.11
BE             164290   0.96
TR             160318   0.93

────────────────────────────────────────────────────────────
TOP 15  STATE  (total unique: 39,888)
────────────────────────────────────────────────────────────
                   count   pct%
state                          
NaN              4958934  28.91
England           719243   4.19
California        619653   3.61
Texas             351371   2.05
New York          309095   1.80
Florida           306345   1.79
So Paulo          240966   1.40
Maharashtra       206449   1.20
Ontario           187025   1.09
le-de-France      185217   1.08
Illinois          165361   0.96
Pennsylvania      141935   0.83
New South Wales   135188   0.79
Georgia           133263   0.78
Massachusetts     129667   0.76

In [10]:
top_cities = value_counts["city"].sort_values(ascending=False).head(20).astype(float)
print(f"\n{'─'*60}")
print("TOP 20 CITIES")
print("─" * 60)
pct = (top_cities / total_rows * 100).round(2)
print(pd.DataFrame({"count": top_cities.astype(int), "pct%": pct}).to_string())



────────────────────────────────────────────────────────────
TOP 20 CITIES
────────────────────────────────────────────────────────────
                 count   pct%
None           3599298  20.98
London          285311   1.66
Paris           133559   0.78
New York        120756   0.70
So Paulo        107430   0.63
Los Angeles      81365   0.47
Madrid           75126   0.44
Toronto          72640   0.42
Dubai            72343   0.42
New Delhi        69718   0.41
Mumbai           68837   0.40
Houston          61732   0.36
Chicago          56732   0.33
San Francisco    51966   0.30
Singapore        51381   0.30
Bangalore        51060   0.30
Berlin           50329   0.29
Barcelona        50027   0.29
Amsterdam        48695   0.28
Sydney           44868   0.26


────────────────────────────────────────────────────────────
TOP 20 CITIES
────────────────────────────────────────────────────────────
                 count   pct%
None           3599298  20.98
London          285311   1.66
Paris           133559   0.78
New York        120756   0.70
So Paulo        107430   0.63
Los Angeles      81365   0.47
Madrid           75126   0.44
Toronto          72640   0.42
Dubai            72343   0.42
New Delhi        69718   0.41
Mumbai           68837   0.40
Houston          61732   0.36
Chicago          56732   0.33
San Francisco    51966   0.30
Singapore        51381   0.30
Bangalore        51060   0.30
Berlin           50329   0.29
Barcelona        50027   0.29
Amsterdam        48695   0.28
Sydney           44868   0.26

In [11]:
if founded_stats:
    arr = np.array(founded_stats)
    # Reasonable range filter for display
    valid = arr[(arr >= 1800) & (arr <= 2026)]
    print(f"\n{'─'*60}")
    print("FOUNDED  (numeric)")
    print("─" * 60)
    print(f"  count     : {len(arr):,}")
    print(f"  nulls     : {total_rows - len(arr):,}  ({(total_rows - len(arr)) / total_rows * 100:.1f}%)")
    print(f"  min       : {arr.min()}")
    print(f"  max       : {arr.max()}")
    print(f"  mean      : {arr.mean():.1f}")
    print(f"  median    : {np.median(arr):.0f}")
    print(f"  std       : {arr.std():.1f}")

    # Decade distribution
    decades = (valid // 10 * 10)
    decade_vc = pd.Series(decades).value_counts().sort_index()
    print(f"\n  Companies by decade founded (valid range 1800-2026):")
    for decade, cnt in decade_vc.items():
        bar = "█" * min(50, int(cnt / decade_vc.max() * 50))
        print(f"  {int(decade)}s  {cnt:>8,}  {bar}")


────────────────────────────────────────────────────────────
FOUNDED  (numeric)
────────────────────────────────────────────────────────────
  count     : 8,503,885
  nulls     : 8,650,132  (50.4%)
  min       : 1200
  max       : 2023
  mean      : 2006.5
  median    : 2013
  std       : 23.6

  Companies by decade founded (valid range 1800-2026):
  1800s       989  
  1810s     1,317  
  1820s     1,693  
  1830s     2,564  
  1840s     3,179  
  1850s     4,954  
  1860s     6,303  
  1870s     7,784  
  1880s    11,187  
  1890s    14,059  
  1900s    19,290  
  1910s    22,799  
  1920s    32,716  
  1930s    31,851  
  1940s    48,297  
  1950s    73,533  
  1960s   113,061  █
  1970s   199,558  ██
  1980s   372,238  ████
  1990s   712,468  █████████
  2000s  1,514,572  ████████████████████
  2010s  3,759,594  ██████████████████████████████████████████████████
  2020s  1,542,354  ████████████████████


────────────────────────────────────────────────────────────
FOUNDED  (numeric)
────────────────────────────────────────────────────────────
  count     : 8,503,885
  nulls     : 8,650,132  (50.4%)
  min       : 1200
  max       : 2023
  mean      : 2006.5
  median    : 2013
  std       : 23.6

  Companies by decade founded (valid range 1800-2026):
  1800s       989  
  1810s     1,317  
  1820s     1,693  
  1830s     2,564  
  1840s     3,179  
  1850s     4,954  
  1860s     6,303  
  1870s     7,784  
  1880s    11,187  
  1890s    14,059  
  1900s    19,290  
  1910s    22,799  
  1920s    32,716  
  1930s    31,851  
  1940s    48,297  
  1950s    73,533  
  1960s   113,061  █
  1970s   199,558  ██
  1980s   372,238  ████
  1990s   712,468  █████████
  2000s  1,514,572  ████████████████████
  2010s  3,759,594  ██████████████████████████████████████████████████
  2020s  1,542,354  ████████████████████

In [ ]:
print(f"\n{'─'*60}")
print("DUPLICATES")
print("─" * 60)
dup_handle = df["handle"].duplicated().sum() if "handle" in df.columns else "N/A"
dup_domain  = df["domain"].duplicated().sum()  if "domain"  in df.columns else "N/A"
dup_full    = df.duplicated().sum()
print(f"  Duplicate handles : {dup_handle:,}")
print(f"  Duplicate domains : {dup_domain:,}")
print(f"  Full-row dupes    : {dup_full:,}")

In [ ]:
if "size" in df.columns and "type" in df.columns:
    print(f"\n{'─'*60}")
    print("CROSS-TAB: size × type  (counts, top values)")
    print("─" * 60)
    ct = pd.crosstab(df["size"], df["type"])
    print(ct.to_string())1


────────────────────────────────────────────────────────────
CROSS-TAB: size × type  (counts, top values)
────────────────────────────────────────────────────────────
type     Educational  Government Agency  Nonprofit  Partnership  Privately Held  Public Company  Self-Employed  Self-Owned
size                                                                                                                      
1-10          127146              14606     475235       673904         2981085          500679         804614     1068133
10K+            5796               4268       5793         2341           10754           15347           1668        1150
11-50         107670              21079     235590       311595         1727850          293590          81950      249706
1K-5K          12389               9120      11155         6604           55334           27881           1562        3573
201-500        23697              13391      31142        32746          210154           5632

────────────────────────────────────────────────────────────
CROSS-TAB: size × type  (counts, top values)
────────────────────────────────────────────────────────────
type     Educational  Government Agency  Nonprofit  Partnership  Privately Held  Public Company  Self-Employed  Self-Owned
size                                                                                                                      
1-10          127146              14606     475235       673904         2981085          500679         804614     1068133
10K+            5796               4268       5793         2341           10754           15347           1668        1150
11-50         107670              21079     235590       311595         1727850          293590          81950      249706
1K-5K          12389               9120      11155         6604           55334           27881           1562        3573
201-500        23697              13391      31142        32746          210154           56320          10382       21644
501-1K         13288               7846      12539        10169           72536           25645           2858        5860
51-200         65246              21140      89201        94060          626258          128002          25583       67848
5K-10K          3608               2419       2428         1293            9289            7842            579         835

## US INDUSTRY EDA